File: flow.c
#include "./include/flow.h"
#include "./include/papi.h"
#include <dlfcn.h>
#include <qhash.h>
#include <stdlib.h>

unsigned class_hd, class_rhd, inst_hd, link_hd, tran_hd, data_hd;
struct flow flow;
unsigned me, parent;
DB_TXN *txnid;

int flow_write(unsigned oport, void *data, size_t size) {
	unsigned key[2] = { me, oport };
	/* fprintf(stderr, "flow_write %u %u\n", me, oport); */
	hash_put(data_hd, key, sizeof(key), data, size);
	return 0;
}

int flow_read(unsigned iport, void *target, size_t size) {
	unsigned key[2] = { parent, iport };
	/* fprintf(stderr, "flow_read %u %u\n", parent, iport); */
	hash_get(data_hd, target, key, sizeof(key));
	return 0;
}

int flow_link_tran(unsigned oinst, unsigned iinst) {
	uhash_put(tran_hd, oinst, &iinst, sizeof(iinst));
	return 0;
}

int flow_link(unsigned oinst, unsigned oport, unsigned iinst, unsigned iport) {
	unsigned key[2], value[2];
	hash_put(link_hd, key, sizeof(key), value, sizeof(value));
	return 0;
}

unsigned flow_node(char *name) {
	static char buf[BUFSIZ];
	unsigned ret;

	sprintf(buf, "./%s.so", name);

	if (!shash_get(class_rhd, &ret, name))
		return ret;

	void *sl = dlopen(buf, RTLD_NOW | RTLD_LOCAL | RTLD_NODELETE);
	if (!sl) {
		fprintf(stderr, "dlopen failed: %s\n", dlerror());
		exit(1);
	}
	ret = lhash_new(class_hd, &sl);
	/* fprintf(stderr, "flow_node new %s %p %u\n", buf, sl, ret); */
	suhash_put(class_rhd, name, ret);

	struct flow *iflow = dlsym(sl, "flow");
	*iflow = flow;

	flow_run_t *flow_init = dlsym(sl, "flow_init");
	if (flow_init)
		(*flow_init)();

	return ret;
}

unsigned flow_inst(unsigned class_id) {
	unsigned ret = lhash_new(inst_hd, &class_id);
	/* fprintf(stderr, "flow_inst %u %u\n", class_id, ret); */
	return ret;
}

void flow_tran_next(unsigned oinst);

unsigned flow_tran(unsigned iinst) {
	void *sl;
	unsigned iclass;
	/* fprintf(stderr, "flow_tran -> %u\n", iinst); */

	uhash_get(inst_hd, &iclass, iinst);
	uhash_get(class_hd, &sl, iclass);
	flow_run_t *flow_run = dlsym(sl, "flow_run");
	me = iinst;
	(*flow_run)();

	flow_tran_next(iinst);
}

void flow_tran_next(unsigned oinst) {
	struct hash_cursor c = hash_iter(tran_hd, &oinst, sizeof(oinst));
	unsigned iinst, iclass;
	/* fprintf(stderr, "flow_tran_next %u\n", oinst); */

	while (hash_next(&oinst, &iinst, &c)) {
		parent = oinst;
		flow_tran(iinst);
	}
}

int main(int argc, char *argv[]) {
	class_hd = lhash_init(sizeof(void *));
	class_rhd = hash_init();
	inst_hd = lhash_init(sizeof(unsigned));
	link_hd = hash_init();
	tran_hd = hash_init();
	data_hd = hash_init();

	flow.read = flow_read;
	flow.write = flow_write;
	flow.link_tran = flow_link_tran;
	flow.link = flow_link;
	flow.node = flow_node;
	flow.inst = flow_inst;
	flow.tran = flow_tran;

	unsigned me_class = flow_node(argv[1]);
	parent = me = flow_inst(me_class);

	flow_tran(me);
	return 0;
}

File: flow.py
import pyflow

class Flow:
    @staticmethod
    def write(port: int, data: bytes) -> int:
        """
        Write into a specified output port.

        Args:
            port (int): The port number.
            data (bytes): The data to write.

        Returns:
            int: The unsigned integer ID of the class.
        """
        buf = pyflow.ffi.new("char[]", data)
        result = self.lib.flow_write(port, buf, len(data))
        return result

    @staticmethod
    def read(port: int, length: int) -> int:
        """
        Read from a specified input port.

        Args:
            port (int): The port number.
            length (int): The maximum number of bytes to read.

        Returns:
            bytes: The data read from the port.
        """
        buf = self.ffi.new("char[]", length)
        read_len = self.lib.flow_read(port, buf, length)
        if read_len < 0:
            raise IOError(f"Failed to read from port {port}, error code {read_len}")
        return self.ffi.buffer(buf, read_len)[:]

    @staticmethod
    def link_tran(oinst: int, iinst: int) -> int:
        """
        Link a transition between nodes

        Args:
            oinst (int): The output node instance
            iinst (int): The input node instance

        Returns:
            int: The return code of the original API
        """
        return int(pyflow.lib.flow_link_tran(oinst, iinst))

    @staticmethod
    def link(oinst: int, oport: int, iinst: int, iport: int) -> int:
        """
        Link a transition between nodes

        Args:
            oinst (int): The output node instance
            oport (int): The output port
            iinst (int): The input node instance
            iport (int): The input port

        Returns:
            int: The return code of the original API
        """
        return int(pyflow.lib.flow_link(oinst, oport, iinst, iport))

    @staticmethod
    def node(name: str) -> int:
        """
        Load a flow or node class.

        Args:
            name (str): The name of the flow node.

        Returns:
            int: The unsigned integer ID of the class.
        """
        c_name = pyflow.ffi.new("char[]", name.encode("utf-8"))
        return int(pyflow.lib.flow_node(c_name))

    @staticmethod
    def inst(class_id: int) -> int:
        """
        Create a new instance of a node / flow.

        Args:
            class_id (int): The ID of the class to create an instance for.

        Returns:
            int: The ID of the instance.
        """
        return int(pyflow.lib.flow_inst(class_id))

    @staticmethod
    def tran(instance_id: int) -> int:
        """
        Transition into specified node instance

        Args:
            instance_id (int): The ID of the instance to transition to.

        Returns:
            int: The original API return value.
        """
        return int(pyflow.lib.flow_tran(instance_id))
